go/contacts/resolution_notice.go

```diff
@@ -0,0 +1,143 @@
+package contacts
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"github.com/keybase/client/go/encrypteddb"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/msgpack"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+type ContactResolution struct {
+	Description  string
+	ResolvedUser keybase1.User
+}
+
+const ContactResolutionEncryptionVersion = 1
+
+type ContactResolutionForEncryption struct {
+	ContactResolution
+	EncryptionVersion int
+}
+
+type EncryptedContactResolution struct {
+	Blob          []byte
+	PukGeneration keybase1.PerUserKeyGeneration
+}
+
+var errWrongContactEncryptionScheme = errors.New(
+	"wrong encryption version for encrypted contact blob")
+
+func getPerUserKeyFn(mctx libkb.MetaContext,
+	pukGen *keybase1.PerUserKeyGeneration) (keyFn func(ctx context.
+	Context) ([32]byte, error), gen keybase1.PerUserKeyGeneration,
+	errOuter error) {
+	pukring, err := mctx.G().GetPerUserKeyring(mctx.Ctx())
+	if err != nil {
+		return nil, gen, err
+	}
+	err = pukring.Sync(mctx)
```

Comment:
what would happen if you're online for a while and the puk rotates on another device, would you need to resync inside keyFn? some tests to make sure it works out after puk roll on both sides would be good